feat: lev aave

fp-crypto opened the pull request.

Commit:
feat: getting started

Commit:
feat: interstitial

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -521,8 +598,74 @@ contract Strategy is BaseStrategy, ICallee {
+        view
+        returns (IAaveIncentivesController)
+    {
+        return aToken.getIncentivesController();
```
IMP as reminder: 
we should be able to avoid calling this function to avoid AAVE removing this method from their implementation and locking the strategy due to calling an inexistent function.

Some aTokens that are not incentivised don't have the function so this function will revert

jmonteer reviewed (commented):
this wip looks good so far

be sure to look at https://github.com/jmonteer/yearnV2-aave-lender-borrower/pull/22
to understand how to deal with Aave

Commit:
feat: interstitial 2

flashfish0x commented:
https://github.com/jmonteer/yearnV2-gen-lev-lending/pull/1/files#diff-3e804d467472e49fe8afea1899d96f26655f3ecd641555a801abcea4f12af598R332

I am thinking we should sell on uni v3 like the badger version does. It smoothes our apr a lot otherwise we will be harvesting loss most of the time. we lose about 2% of rewards but i think it is worth it for the smoothing and hedging against price crash.

See here:
https://github.com/GalloDaSballo/wbtc-levered-strat/blob/master/contracts/Strategy.sol#L352

There is a mistake in the fees they chose but apart from that it is just copy and paste

fp-crypto commented:
> https://github.com/jmonteer/yearnV2-gen-lev-lending/pull/1/files#diff-3e804d467472e49fe8afea1899d96f26655f3ecd641555a801abcea4f12af598R332
> 
> I am thinking we should sell on uni v3 like the badger version does. It smoothes our apr a lot otherwise we will be harvesting loss most of the time. we lose about 2% of rewards but i think it is worth it for the smoothing and hedging against price crash.
> 
> See here:
> https://github.com/GalloDaSballo/wbtc-levered-strat/blob/master/contracts/Strategy.sol#L352
> 
> There is a mistake in the fees they chose but apart from that it is just copy and paste

Sounds good

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -32,24 +36,36 @@ contract Strategy is BaseStrategy, ICallee {
+    // Aave's referral code
+    uint16 internal referral;
+
     uint256 private a = 0;
```
what is a?

jmonteer replied:
is it from the lev comp template?

fp-crypto replied:
yup 🤣. i was wondering the same thing

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -119,36 +141,44 @@ contract Strategy is BaseStrategy, ICallee {
+        if (deposits == 0 || totalDeposits == 0) {
             return 0;
         }
         // NOTE: compSpeed is the same for borrows and deposits
```
obsolete comment

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -161,13 +191,10 @@ contract Strategy is BaseStrategy, ICallee {
-        // update position
-        getUpdatedPosition();
-        
+
```
https://github.com/fp-crypto/yearnV2-gen-lev-lending/blob/171b62d40003c5c6a44045f8360f0bf339c98bd6/contracts/Strategy.sol#L205-L206

probably better to call `_claimAndSellRewards()` before calling `estimatedTotalAssets()` so we get the real value of rewards (which should be pretty close as we are using uniswap to estimate but just in case)

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -230,13 +260,12 @@ contract Strategy is BaseStrategy, ICallee {
-
         } else if (currentCollatRatio < targetCollatRatio) {
             // we should lever up
             // TODO: create minRatio state variable
```
https://github.com/fp-crypto/yearnV2-gen-lev-lending/blob/171b62d40003c5c6a44045f8360f0bf339c98bd6/contracts/Strategy.sol#L271

ooops forgot to delete this

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -291,40 +322,65 @@ contract Strategy is BaseStrategy, ICallee {
+        assets[0] = address(aToken);
+        assets[1] = address(debtToken);
+
+        _incentivesController().claimRewards(
```
https://github.com/fp-crypto/yearnV2-gen-lev-lending/blob/171b62d40003c5c6a44045f8360f0bf339c98bd6/contracts/Strategy.sol#L350

could pose problems for those tokens that have not incentives OR if incentives are removed!

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -291,40 +322,65 @@ contract Strategy is BaseStrategy, ICallee {
+        uint256 UNSTAKE_WINDOW = stkAave.UNSTAKE_WINDOW();
+        if (
+            (IERC20(address(stkAave)).balanceOf(address(this)) > 0 &&
+                (cooldownStartTimestamp == 0)) ||
```
https://github.com/fp-crypto/yearnV2-gen-lev-lending/blob/171b62d40003c5c6a44045f8360f0bf339c98bd6/contracts/Strategy.sol#L362-L365

parenthesis should be the other way around

IF there is balance of stkAaave AND (cooldown == 0 OR block.timestamp > ...) 

you always need positive balance iirc

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -291,40 +322,65 @@ contract Strategy is BaseStrategy, ICallee {
         (uint256 deposits, uint256 borrows, ) = getCurrentPosition();
 
-        // NOTE: we cannot 
+        // NOTE: we cannot
```
https://github.com/fp-crypto/yearnV2-gen-lev-lending/blob/171b62d40003c5c6a44045f8360f0bf339c98bd6/contracts/Strategy.sol#L376

can't remember what I was referring to with this... sorry

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -470,49 +560,57 @@ contract Strategy is BaseStrategy, ICallee {
+    {
+        borrows = balanceOfAToken();
+        deposits = balanceOfDebtToken();
+        if (deposits > 0) {
```
https://github.com/fp-crypto/yearnV2-gen-lev-lending/blob/171b62d40003c5c6a44045f8360f0bf339c98bd6/contracts/Strategy.sol#L578

warning: 
not all aTokens have 18 decimals

aWBTC has 8 decimals!

jmonteer replied:
use DECIMALS

jmonteer reviewed (commented):
good job! looks pretty closed to being finished. 

suggestions: 
- add a way to sell rewards directly using v3 liquidity stkAave -> aave -> eth -> want instead of waiting for the cooldown
- add manual method to sell rewards
- add manual methods for lever down, lever up, withdraw collateral, repay debt, ...

Commit:
feat: further along

Commit:
feat: more working

Commit:
feat: more work, maybe closer?

Commit:
feat: further

Commit:
feat: pretty much down?

Commit:
feat: mostly done

Commit:
feat: cleanup

Commit:
feat: working tend

Commit:
feat: clean up

Commit:
feat: additional setters

Commit:
feat: more cleanup

Commit:
chore: slightly smaller bytecode size

Commit:
fix: approve oopsies

Commit:
fix: sellStkAave

Commit:
fix: tests

Commit:
chore: missing files

flashfish0x commented on contracts/Strategy.sol:
```diff
@@ -264,18 +390,43 @@ contract Strategy is BaseStrategy, ICallee {
+        (_amountFreed, ) = liquidatePosition(type(uint256).max);
     }
 
     function prepareMigration(address _newStrategy) internal override {
```
instead of leaving this empty I'd do a revert is balance (not including want) is more than dustthreshold. That way we cant revert and leave funds unles we manually increase dustthreshold loads

fp-crypto replied:
done

flashfish0x commented on contracts/Strategy.sol:
```diff
@@ -461,68 +677,231 @@ contract Strategy is BaseStrategy, ICallee {
+            abi.decode(data, (bool, uint256, address));
         require(msg.sender == FlashLoanLib.SOLO);
         require(sender == address(this));
+        require(token == address(want));
```
why do you have this line? why is token passed in when it is always want?

fp-crypto replied:
good point

flashfish0x commented on contracts/Strategy.sol:
```diff
@@ -461,68 +677,231 @@ contract Strategy is BaseStrategy, ICallee {
+            bytes memory path =
+                abi.encodePacked(
+                    address(aave),
+                    uint24(10000),
```
is this the correct fee? it looks wrong

flashfish0x replied:
definitely not correct. 0.3% pool is best for aave-> eth
weth-> want probably varies. 

Maybe make the path settable? hardcoding here could rekt us

flashfish0x commented on contracts/Strategy.sol:
```diff
@@ -461,68 +677,231 @@ contract Strategy is BaseStrategy, ICallee {
+            ISwapRouter.ExactInputSingleParams(
+                address(stkAave),
+                address(aave),
+                10000,
```
this fee is too high as well

flashfish0x replied:
0.3% looks better

fp-crypto replied:
1% is better here, see: https://info.uniswap.org/#/tokens/0x4da27a545c0c5b758a6ba100e3a049001de870f5
<img width="1267" alt="Screen Shot 2021-09-09 at 6 30 13 PM" src="https://user-images.githubusercontent.com/83050944/132784226-7ac01b61-47dc-4111-bd68-bcede26baef7.png">

fp-crypto replied:
regardless, i've made this configurable

flashfish0x reviewed (commented):
looks good mostly!

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -19,136 +19,245 @@ import {
+    function _initializeThis() internal {
+        require(address(aToken) == address(0));
+
+        //healthCheck = address(0xDDCea799fF1699e98EDF118e0629A974Df7DF012); // health.ychad.eth
```
obsolete?

fp-crypto replied:
i need to uncomment before deploy, it makes testing more painful

fp-crypto replied:
Actually, it expands the bytecode, we'll just need to set it after deploy.

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -461,68 +677,231 @@ contract Strategy is BaseStrategy, ICallee {
+        uint256 _amount,
+        uint256 _seconds
+    ) private view returns (uint256) {
+        (, uint256 aavePerSecond, ) = incentivesController.getAssetData(_token);
```
what happens if incentives controller does not have a registry for this token? error? or 0 value

fp-crypto replied:
I have not seen this case for any of the tokens we care about. It will return 0 values if it's not registered.

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -19,136 +19,245 @@ import {
+    // AAVE protocol address
+    IProtocolDataProvider private constant protocolDataProvider =
+        IProtocolDataProvider(0x057835Ad21a177dbdd3090bB1CAE03EaCF78Fc6d);
+    IAaveIncentivesController private constant incentivesController =
```
is this safe to hardcode? can it change?

fp-crypto replied:
It's a proxy, so i believe it's safe.

fp-crypto replied:
See: https://etherscan.io/address/0xd784927Ff2f95ba542BfC824c8a8a98F3495f6b5#code

fp-crypto replied:
<img width="656" alt="Screen Shot 2021-09-09 at 7 45 46 PM" src="https://user-images.githubusercontent.com/83050944/132791136-94ae9ef9-5acb-46f1-91a5-dc8aeda75fbe.png">
attempting to confirm with the aave team

fp-crypto replied:
<img width="1164" alt="Screen Shot 2021-09-09 at 8 43 22 PM" src="https://user-images.githubusercontent.com/83050944/132796290-194ff48c-2d2f-48a9-b6f3-e448d2a42238.png">
Aave team member says it's safu

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -160,42 +269,44 @@ contract Strategy is BaseStrategy, ICallee {
+
+        // Assets immediately convertable to want only
+        uint256 supply = getCurrentSupply();
+        uint256 totalAssets = balanceOfWant().add(supply);
```
it should be the same but why not using estimatedTotalAssets? 

i'd rather standarize this

jmonteer replied:
also, (very slightly) lower bytecode :D

fp-crypto replied:
estimatedTotalAssets includes rewards

jmonteer replied:
which are 0 here bc you just sold them right?

fp-crypto replied:
> which are 0 here bc you just sold them right?

could be cooling them down

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -290,41 +441,74 @@ contract Strategy is BaseStrategy, ICallee {
+        (uint256 deposits, uint256 borrows) = getCurrentPosition();
 
-        // NOTE: we cannot 
+        // NOTE: we cannot
```
remove comment. still cant remember why i put this

Commit:
chore: fishy review

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -333,126 +517,158 @@ contract Strategy is BaseStrategy, ICallee {
-        if(amount == 0) return 0;
-        require(cToken.mint(amount) == 0, "!mint");
+        if (amount == 0) return 0;
+        lendingPool.deposit(address(want), amount, address(this), referral);
```
can we not have enough balance? what happens then?

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -290,41 +441,74 @@ contract Strategy is BaseStrategy, ICallee {
-            return amounts[amounts.length - 1];
+    function _claimAndSellRewards() internal returns (uint256) {
+        uint256 stkAaveBalance = balanceOfStkAave();
+        uint8 cooldownStatus = stkAaveBalance == 0 ? 0 : _checkCooldown(); // don't check status if we have no stkAave
```
why uint8 instead of bool?

fp-crypto replied:
This allows me to differentiate between 3 states:
> 0. no cooldown initiated
> 1. active claim period
> 2. future claim period

jmonteer reviewed (commented):
LGTM

Commit:
fix: missing links

Commit:
fix: under bytesize limit, barely

Commit:
chore: black tests

Commit:
fix: regression

Commit:
fix: higher price impact so tests don't fail

Commit:
chore: cleanup

Commit:
chore: comments

Commit:
feat: cooldown test

Commit:
feat: minor gas savings

Commit:
fix: address @salazarguille's comments

https://github.com/yearn/yearn-strategies/issues/136#issuecomment-918557976
https://github.com/yearn/yearn-strategies/issues/136#issuecomment-918576038
https://github.com/yearn/yearn-strategies/issues/136#issuecomment-918577312
https://github.com/yearn/yearn-strategies/issues/136#issuecomment-918631019
https://github.com/yearn/yearn-strategies/issues/136#issuecomment-918634162

Commit:
fix: more test coverage

Commit:
feat: improve estimatedRewards implementation & more setting checks

https://github.com/yearn/yearn-strategies/issues/136#issuecomment-919386382

Commit:
fix: don't include rewards in estimatedAssets there is no position open

Commit:
chore: simplify healthcheck tests

Commit:
feat: more deleverage tests

Commit:
chore: test improvement

Commit:
fix: bug in setCollat
